File: examples/dynamic_data_sequences/java/DynamicDataSequences.java
import com.rti.dds.dynamicdata.DynamicData;
import com.rti.dds.dynamicdata.DynamicDataInfo;
import com.rti.dds.dynamicdata.DynamicDataMemberInfo;
import com.rti.dds.typecode.*;

/*********
 * IDL representation for this example ************
 * 
 * public class SimpleStruct { long a_member; };
 * 
 * public class TypeWithSequence { sequence<256,SimpleStruct> sequence_member;
 * };
 */

/*
 * Uncomment to use the bind api instead of the get API.
 */
// Constants.bin_api = true;

public class DynamicDataSequences {
    public static int MAX_SEQ_LEN = 5;
    public static boolean bin_api = false;

    /*
     * Uncomment to use the bind api instead of the get API.
     */
    // bin_api = true;

    static TypeCode sequence_element_get_typecode() {
        StructMember members[] = new StructMember[0];
        TypeCode tc;

        /* First, we create the typeCode for the simple struct */
        try {
            tc = TypeCodeFactory.TheTypeCodeFactory.create_struct_tc("Foo",
                    members);
        } catch (Exception e) {
            System.err.println(e.getMessage());
            return null;
        }

        try {
            tc.add_member("a_member", TypeCode.MEMBER_ID_INVALID,
                    TypeCode.TC_LONG, TypeCode.NONKEY_MEMBER);
        } catch (Exception e) {
            System.err.println(e.getMessage());
            return null;
        }

        return tc;

    }

    static TypeCode sequence_get_typecode() {
        TypeCode tc;

        TypeCode seqElementTC = sequence_element_get_typecode();
        if (seqElementTC == null) {
            System.err.println("! Unable to create TypeCode");
            return null;
        }

        /* We create the typeCode for the sequence */
        try {
            tc = TypeCodeFactory.TheTypeCodeFactory.create_sequence_tc(
                    MAX_SEQ_LEN, seqElementTC);
        } catch (Exception e) {
            System.err.println(e.getMessage());
            return null;
        }

        if (seqElementTC != null) {
            TypeCodeFactory.TheTypeCodeFactory.delete_tc(seqElementTC);
        }

        return tc;

    }

    static TypeCode type_w_sequence_get_typecode() {
        StructMember members[] = new StructMember[0];
        
        TypeCode sequenceTC = sequence_get_typecode();
        if (sequenceTC == null) {
            System.err.println("! Unable to create Type with sequence TC");
            return null;
        }

        TypeCode tc = TypeCodeFactory.TheTypeCodeFactory.create_struct_tc(
                "TypeWithSequence", null);
        if (tc == null) {
            System.err.println("! Unable to create Type with sequence TC");
        }

        try {
            tc.add_member("sequence_member", TypeCode.MEMBER_ID_INVALID,
                    sequenceTC, TypeCode.NONKEY_MEMBER);
        } catch (Exception e) {
            System.err.println(e.getMessage());
            return null;
        }

        if (sequenceTC != null) {
            TypeCodeFactory.TheTypeCodeFactory.delete_tc(sequenceTC);
        }

        return tc;

    }

    public static void write_data(DynamicData sample) {
        /* Creating TypeCodes */
        TypeCode sequenceTC = sequence_get_typecode();
        if (sequenceTC == null) {
            System.err.println("! Unable to create sequence typecode");
            return;
        }

        TypeCode sequenceElementTC = sequence_element_get_typecode();
        if (sequenceElementTC == null) {
            System.err.println("! Unable to create sequence element TypeCode");
            return;
        }

        DynamicData seqmember = new DynamicData(sequenceTC,
                DynamicData.PROPERTY_DEFAULT);
        DynamicData seqelement = new DynamicData(sequenceElementTC,
                DynamicData.PROPERTY_DEFAULT);

        for (int i = 0; i < MAX_SEQ_LEN; ++i) {
            /*
             * To access the elements of a sequence it is necessary to use their
             * id. This parameter allows accessing to every element of the
             * sequence using a 1-based index. There are two ways of doing this:
             * bind API and get API. See the NestedStructExample for further
             * details about the differences between these two APIs.
             */

            if (bin_api == true) { // bind API
                try {
                    seqmember.bind_complex_member(seqelement, null, i + 1);
                    seqelement.set_long("a_member",
                            DynamicData.MEMBER_ID_UNSPECIFIED, i);
                    System.out.println("Writing sequence element #" + (i + 1)
                            + " : ");
                    System.out.println(seqelement);
                } catch (Exception e) {
                    System.err.println(e.getMessage());
                }
            } else { // get API
                try {
                    seqelement.set_long("a_member",
                            DynamicData.MEMBER_ID_UNSPECIFIED, i);
                    System.out.println("Writing sequence element #" + (i + 1)
                            + " : ");
                    seqmember.set_complex_member(null, i + 1, seqelement);
                } catch (Exception e) {
                    System.err.println(e.getMessage());
                }
            }

        }

        try {
            sample.set_complex_member("sequence_member",
                    DynamicData.MEMBER_ID_UNSPECIFIED, seqmember);
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }

        if (sequenceTC != null) {
            TypeCodeFactory.TheTypeCodeFactory.delete_tc(sequenceTC);
        }

        if (sequenceElementTC != null) {
            TypeCodeFactory.TheTypeCodeFactory.delete_tc(sequenceElementTC);
        }

        return;

    }

    public static void read_data(DynamicData sample) {
        TypeCode sequenceTC = sequence_get_typecode();
        if (sequenceTC == null) {
            System.err.println("! Unable to get sequence Typecode");
            return;
        }

        DynamicData seqmember = new DynamicData(sequenceTC,
                DynamicData.PROPERTY_DEFAULT);
        DynamicData seqelement = new DynamicData(null,
                DynamicData.PROPERTY_DEFAULT);

        long value = 1;
        int seqlen = 0;
        DynamicDataInfo info = new DynamicDataInfo();

        sample.get_complex_member(seqmember, "sequence_member",
                DynamicData.MEMBER_ID_UNSPECIFIED);

        /*
         * Now we get the total amount of elements contained in the array by
         * accessing the dynamic data info
         */
        System.out.println("* Getting sequence member info....");
        seqmember.get_info(info);

        seqlen = info.member_count;
        System.out.println("* Sequence contains " + seqlen + " elements");

        for (int i = 0; i < seqlen; ++i) {
            /*
             * The same results can be obtained using bind_complex_member
             * method. The main difference, is that in that case the members are
             * not copied, just referenced
             */
            if (bin_api) { // using bind API
                try {
                    seqmember.bind_complex_member(seqelement, null, i + 1);
                } catch (Exception e) {
                    System.out.println(e.getMessage());
                }
            } else { // using get API
                try {
                    seqmember.get_complex_member(seqelement, null, i + 1);
                } catch (Exception e) {
                    System.err.println(e.getMessage());
                }
            }

            value = seqmember.get_long("a_member",
                    DynamicData.MEMBER_ID_UNSPECIFIED);
            System.out.println("Readeing sequence element #" + (i + 1) + " :");
            System.out.println(seqelement);
            
            if (bin_api) {
                try {
                    seqmember.unbind_complex_member(seqelement);
                } catch (Exception e) {
                    System.out.println(e.getMessage());
                }
            }
        }

    }

    public static void main(String[] args) {

        TypeCode wSequenceTC = type_w_sequence_get_typecode();
        if (wSequenceTC == null) {
            System.err.println("! Unable to create wSequence TypeCode");
            return;
        }

        DynamicData sample = new DynamicData(wSequenceTC,
                DynamicData.PROPERTY_DEFAULT);
        System.out.println("***** Writing a sample *****");
        write_data(sample);

        System.out.println("***** Reading a sample *****");
        read_data(sample);

        /* Delete the created TC */
        if (wSequenceTC != null) {
            TypeCodeFactory.TheTypeCodeFactory.delete_tc(wSequenceTC);
        }
    }
}

File: examples/dynamic_data_sequences/cs/dynamic_data_sequences.cs
/*******************************************************************************
 (c) 2005-2014 Copyright, Real-Time Innovations, Inc.  All rights reserved.
 RTI grants Licensee a license to use, modify, compile, and create derivative
 works of the Software.  Licensee has the right to distribute object form only
 for use with RTI products.  The Software is provided "as is", with no warranty
 of any type, including any warranty for fitness for any purpose. RTI is under
 no obligation to maintain or support the Software.  RTI shall not be liable for
 any incidental or consequential damages arising out of the use or inability to
 use the software.
 ******************************************************************************/
/*
 * Uncomment to use the bind api instead of the get API.
 */
//#define USE_BIND_API 

using System;
using System.Collections.Generic;
using System.Text;
/* dynamic_data_sequences.cs

  This example
  - creates the type code of a sequence
  - creates a DynamicData instance
  - writes and reads the values

 */

/********* IDL representation for this example ************
 *
 * struct SimpleStruct {
 *     long a_member;
 * };
 *
 * struct TypeWithSequence {
 * 	   sequence<SimpleStruct,5> sequence_member;
 * };
 */


public class sequencesExample {
    private static uint MAX_SEQ_LEN = 5;
    private const String sequence_member_name = "sequence_member";
    private static DDS.TypeCode sequence_element_get_typecode(
        DDS.TypeCodeFactory tcf) {
        DDS.StructMemberSeq members = new DDS.StructMemberSeq();

        try {
            /* First, we create the typeCode for the simple struct */
            DDS.TypeCode tc = tcf.create_struct_tc("SimpleStruct", members);

            tc.add_member("a_member", DDS.TypeCode.MEMBER_ID_INVALID,
                    DDS.TypeCode.TC_LONG, DDS.TypeCode.NONKEY_MEMBER);

            return tc;
        } catch (DDS.Exception e) {
            Console.WriteLine("register_type error {0}", e);
            return null;
        }

    }

    private static DDS.TypeCode sequence_get_typecode(
        DDS.TypeCodeFactory tcf ) {
       
        try {
            DDS.TypeCode seq_element_tc = sequence_element_get_typecode(tcf);
            if (seq_element_tc == null) {
                Console.WriteLine("Error to create " +
                    "sequence_element_get_typecode");
                return null;
            }
            /* We create the typeCode for the sequence */
            DDS.TypeCode tc = tcf.create_sequence_tc(MAX_SEQ_LEN, 
                    seq_element_tc);

            return tc;
        } catch (DDS.Exception e) {
            Console.WriteLine("register_type error {0}", e);
            return null;
        }

    }

    private static DDS.TypeCode type_w_sequence_get_typecode(
    DDS.TypeCodeFactory tcf ) {
        DDS.StructMemberSeq members = new DDS.StructMemberSeq();

        try {
            DDS.TypeCode sequence_tc = sequence_get_typecode(tcf);
            if (sequence_tc == null) {
                Console.WriteLine("Error to create sequence_get_typecode");
                return null;
            }
            /* We create the typeCode for struct which contains the sequence */
            DDS.TypeCode tc = tcf.create_struct_tc("TypeWithSequence",
                members);

            tc.add_member(sequence_member_name, DDS.TypeCode.MEMBER_ID_INVALID,
                sequence_tc, DDS.TypeCode.NONKEY_MEMBER);

            return tc;
        } catch (DDS.Exception e) {
            Console.WriteLine("register_type error {0}", e);
            return null;
        }

    }

    private static void write_data (DDS.DynamicData sample, 
        DDS.TypeCodeFactory tcf) {
        /* Creating typecodes */
        DDS.TypeCode sequence_tc = sequence_get_typecode(tcf);
        if (sequence_tc == null) {
            Console.WriteLine("Error to create sequence_get_typecode in " +
                "writing_data");
            return;
        }

        DDS.TypeCode sequence_element_tc = sequence_element_get_typecode(tcf);
        if (sequence_element_tc == null) {
            Console.WriteLine("Error to create sequence_element_get_typecode " +
                "in writing_data");
            return;
        }

        try {
            /* Creating DynamicData */
            DDS.DynamicData seq_member = new DDS.DynamicData(sequence_tc,
                DDS.DynamicData.DYNAMIC_DATA_PROPERTY_DEFAULT);
            DDS.DynamicData seq_element = 
                new DDS.DynamicData(sequence_element_tc,
                DDS.DynamicData.DYNAMIC_DATA_PROPERTY_DEFAULT);

            for (int i = 0; i < MAX_SEQ_LEN; ++i) {
                /* To access the elements of a sequence it is necessary
                 * to use their id. This parameter allows accessing to every 
                 * element of the sequence using a 1-based index.
                 * There are two ways of doing this: bind API and get API.
                 * See the NestedStructExample for further details about the 
                 * differences between these two APIs. */

#if (USE_BIND_API)
                seq_member.bind_complex_member(seq_element, null, i + 1);
                seq_element.set_int("a_member", 
                    DDS.DynamicData.MEMBER_ID_UNSPECIFIED, i);
                Console.WriteLine("Writing sequence elemente #{0} : ", i + 1);
                seq_element.print(1);
                seq_member.unbind_complex_member(seq_element);
#else
                seq_element.set_int("a_member",
                    DDS.DynamicData.MEMBER_ID_UNSPECIFIED, i);
                Console.WriteLine("Writing sequence element #{0}", i + 1);
                seq_element.print(1);
                seq_member.set_complex_member(null, i + 1, seq_element);
#endif
            }

            sample.set_complex_member(sequence_member_name,
                DDS.DynamicData.MEMBER_ID_UNSPECIFIED, seq_member);

            /* Delete the created TC */
            if (sequence_element_tc != null) {
                tcf.delete_tc(sequence_element_tc);
            }

            if (sequence_tc != null) {
                tcf.delete_tc(sequence_tc);
            }
            
            return;
        } catch (DDS.Exception e) {
            Console.WriteLine("register_type error {0}", e);
            return;
        }
    }

    private static void read_data(DDS.DynamicData sample, 
        DDS.TypeCodeFactory tcf ) {
       
        /* Creating typecodes */
        DDS.TypeCode sequence_tc = sequence_get_typecode(tcf);
        if (sequence_tc == null) {
            Console.WriteLine("Error to create sequence_get_typecode in " + 
                "reading_data");
            return;
        }

        /* Creating DynamicData */
        DDS.DynamicData seq_member = new DDS.DynamicData(sequence_tc,
            DDS.DynamicData.DYNAMIC_DATA_PROPERTY_DEFAULT);
        DDS.DynamicData seq_element = new DDS.DynamicData(null,
            DDS.DynamicData.DYNAMIC_DATA_PROPERTY_DEFAULT);
        
        DDS.DynamicDataInfo info = new DDS.DynamicDataInfo();

        sample.get_complex_member(seq_member, sequence_member_name, 
            DDS.DynamicData.MEMBER_ID_UNSPECIFIED);
            
        /* Now we get the total amount of elements contained in the array
         * by accessing the dynamic data info
         */
        Console.WriteLine("* Getting sequence member info....");
        seq_member.get_info(info);
        Console.WriteLine("* Sequence contains {0} elements", 
            info.member_count);

        for (int i = 0; i < info.member_count; ++i) {
            /*
             * The same results can be obtained using
             * bind_complex_member method. The main difference, is that
             * in that case the members are not copied, just referenced
             */

#if (USE_BIND_API)
             try {
                 seq_member.bind_complex_member(seq_element, null, i + 1);
             } catch (DDS.Exception e) {
                 Console.WriteLine("register_type error {0}", e);
                 return;
             }

#else
            try {
                seq_member.get_complex_member(seq_element, null, i + 1);
            } catch (DDS.Exception e) {
                Console.WriteLine("register_type error {0}", e);
                return;
            }
            
#endif
            long value = seq_element.get_long("a_member", 
                DDS.DynamicData.MEMBER_ID_UNSPECIFIED);
            Console.WriteLine("Reading sequence element #{0} : ",
                i + 1);
            seq_element.print(1);

#if (USE_BIND_API)
            try {
                seq_member.unbind_complex_member(seq_element);
            } catch (DDS.Exception e) {
                Console.WriteLine("register_type error {0}", e);
                return;
            }
#endif
        }

        /* Delete the created TC */
        if (sequence_tc != null) {
            tcf.delete_tc(sequence_tc);
        }

        return;
    }

    public static void Main(string[] args) {
       DDS.TypeCodeFactory tcf = DDS.TypeCodeFactory.get_instance();

        DDS.TypeCode w_sequence_tc = type_w_sequence_get_typecode(tcf);
        if (w_sequence_tc == null) {
            Console.WriteLine("Error to create type_w_sequence_get_typecode");
            return;
        }
        try {
            DDS.DynamicData sample = new DDS.DynamicData(w_sequence_tc,
                DDS.DynamicData.DYNAMIC_DATA_PROPERTY_DEFAULT);

            Console.WriteLine("***** Writing a sample *****");
            write_data(sample, tcf);

            Console.WriteLine("***** Reading a sample *****");
            read_data(sample, tcf);

            /* Delete the created TC */
            if (w_sequence_tc != null) {
                tcf.delete_tc(w_sequence_tc);
            }
        } catch (DDS.Exception e) {
            Console.WriteLine("register_type error {0}", e);
            return;
        }
    }
}